src/new-storage/BucketManager.js

```diff
@@ -0,0 +1,336 @@
+        } else if (fromTimestamp !== undefined && toTimestamp === undefined) {
+            query = GET_LAST_BUCKETS_FROM_TIMESTAMP
+            params = [streamId, partition, fromTimestamp]
+        } else if (fromTimestamp === undefined && toTimestamp !== undefined) {
+            query = GET_LAST_BUCKETS_TO_TIMESTAMP
+            params = [streamId, partition, toTimestamp]
+        } else {
+            throw TypeError(`Not correct combination of fromTimestamp (${fromTimestamp}) and toTimestamp (${toTimestamp})`)
+        }
+
+        return this._getBucketsFromDatabase(query, params, streamId, partition)
+    }
+
+    /**
+     * Get latest N buckets or get latest N buckets before some date (to check buckets in the past)
+     *
+     * @param streamId
+     * @param partition
+     * @param limit
+     * @param timestamp
+     * @returns {Promise<[]>}
+     */
+    async getLastBuckets(streamId, partition, limit = 1, timestamp = undefined) {
+        const GET_LAST_BUCKETS = 'SELECT * FROM bucket WHERE stream_id = ? and partition = ? LIMIT ?'
+        const GET_LAST_BUCKETS_TIMESTAMP = 'SELECT * FROM bucket WHERE stream_id = ? and partition = ? AND date_create <= ? LIMIT ?'
+
+        let query
+        let params
+
+        if (timestamp) {
+            query = GET_LAST_BUCKETS_TIMESTAMP
+            params = [streamId, partition, timestamp, limit]
+        } else {
+            query = GET_LAST_BUCKETS
+            params = [streamId, partition, limit]
+        }
+
+        return this._getBucketsFromDatabase(query, params, streamId, partition)
+    }
+
+    async _getBucketsFromDatabase(query, params, streamId, partition) {
+        const buckets = []
+
+        try {
+            const resultSet = await this.cassandraClient.execute(query, params, {
+                prepare: true,
+            })
+
+            resultSet.rows.forEach((row) => {
+                const { id, records, size, date_create: dateCreate } = row
+
+                const bucket = new Bucket(
+                    id.toString(), streamId, partition, size, records, new Date(dateCreate),
+                    this.opts.maxBucketSize, this.opts.maxBucketRecords, this.opts.bucketKeepAliveSeconds
+                )
+
+                buckets.push(bucket)
+            })
+        } catch (e) {
+            if (this.opts.logErrors) {
+                console.error(e)
+            }
+        }
+
+        return buckets
+    }
+
+    stop() {
+        clearInterval(this._checkFullBucketsTimeout)
+        clearInterval(this._storeBucketsTimeout)
+    }
+
+    _storeBuckets() {
+        // for non-existing buckets UPDATE works as INSERT
+        const UPDATE_BUCKET = 'UPDATE bucket SET size = ?, records = ?, id = ? WHERE stream_id = ? AND partition = ? AND date_create = ?'
+
+        const notStoredBuckets = Object.values(this.buckets).filter((bucket) => !bucket.isStored())
+
+        notStoredBuckets.forEach(async (bucket) => {
+            bucket.setStored()
```

Comment:
I added check `if (bucket.records === records)` and now I remove buckets from memory only if `if (!bucket.isAlive() && bucket.isStored())`.

Don't fully understand what you mean `on next setTimeout though if time taken by execute is longer than setTimeout`. 

I added example, which describes `_storeBuckets`, you can run locally:
```js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
}

async function storeToDb(i) {
    console.log(`${i}\n`)
    await sleep(1000)
}

async function main() {
    for (let i = 0; i < 3; i++) {
        await storeToDb(i)
    }
    console.log('next iteration\n')
    setTimeout(() => main(), 5000)
}

main()
``` 

So how it works. When we call `_storeBuckets`, we loop through all buckets and store them. `_storeBuckets` is not called until all buckets are `inserted` or `failed` to insert, because we await for each `execute`. And only after that we call again `setTimeout(() => this._storeBuckets(), this.opts.storeBucketsTimeout)`

Reply:
Yeah your code snippet is what I'm describing. It waits correctly with `for` but not with `forEach`.
Compare:
```js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
}

async function storeToDb(i) {
    await sleep(Math.random() * 4000)
    console.info(`executed ${i}`)
}

async function main() {
    console.info('-- start main')
    for (let i = 0; i < 3; i++) {
        await storeToDb(i)
    }

    console.info('-- end main')
    setTimeout(() => main(), 2000)
}

main()
``` 

vs.

```js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
}

async function storeToDb(i) {
    await sleep(Math.random() * 4000)
    console.info(`executed ${i}`)
}

async function main() {
    console.info('-- start main')
    const rounds = [0, 1, 2]
    rounds.forEach(async (i) => {
        await storeToDb(i)
    })

    console.info('-- end main')
    setTimeout(() => main(), 2000)
}

main()
```

I think you need to add `Promise.allSettled()`:

```js
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
}

async function storeToDb(i) {
    await sleep(Math.random() * 4000)
    console.info(`executed ${i}`)
}

async function main() {
    console.info('-- start main')
    const rounds = [0, 1, 2]
    await Promise.allSettled(rounds.map(async (i) => {
        await storeToDb(i)
    }))

    console.info('-- end main')
    setTimeout(() => main(), 2000)
}

main()
```